### PR TITLE
An n-pass linter

### LINKED ISSUE #12: Change to 2-pass lint process
To handle fA() calling fB() before fB() is defined.

### PR DESCRIPTION
**Pull request by TotalVerb**
This is a big change that switches the current one-pass linter to a as-many-passes-as-needed linter. It should be a big step toward removing spurious symbol not found problems. Currently there are a lot of broken tests, but they should mostly not be too hard to fix.

This breaks a lot of backward compatibility for `LintContext`, `LintMessage`, etc., because large portions of the infrastructure had to be rewritten. I'll document the changes.

Closes #12.

### CODE REVIEW

**Comment by TotalVerb** [comment]
Running this on packages tends to yield a barrage of "local variable shadows global" problems (the new I340, I341, I342, I343). This is consistent with every package I try it on. They're not false positives but very annoying, so it might make sense to disable them by default.

File: src/result.jl [hunk 1]
```diff
@@ -1,7 +1,9 @@
+# TODO: when 0.5 support dropped, remove the isdefined check
 const LINT_RESULT_COLORS = Dict(
-    :INFO => :bold,
-    :WARN => :yellow,
-    :ERROR => :magenta,
+    :INFO => Base.info_color(),
+    :WARN => Base.warn_color(),
+    :ERROR => isdefined(Base, :error_color) ? Base.error_color() :
```
Comment (musm) [inline src/result.jl]: would it ever be the case where this is not defined ?

Comment (TotalVerb) [inline src/result.jl]: On 0.5, `Base.error_color` doesn't exist, because errors and warnings are both printed in `Base.warn_color()`.

**Comment by TotalVerb** [comment]
I think I'll merge this soon to give it time to simmer on master, in case people who are `checkout`ing Lint are depending on the removed features. However this isn't ready for tagging yet; there are still a lot of broken tests that need to be fixed.

**Pull request merged**